monitor.go

```diff
@@ -51,50 +51,63 @@ func (a *autographer) handleMonitor(w http.ResponseWriter, r *http.Request) {
 		go func(i int, s signer.Signer) {
 			defer wg.Done()
 
-			if s.Config().Type == apk2.Type {
-				// apk2 signer cannot be monitored because it only
-				// works with full APK files, which is too heavy for this
-				// monitoring function, so we return a signature response
-				// without a signature, only metadata
+			// First try the DataSigner interface. If the signer doesn't
+			// implement it, try the FileSigner interface. If that's still
+			// not implemented, return an error.S
+			if _, ok := s.(signer.DataSigner); ok {
+				// sign with data set to the base64 of the string 'AUTOGRAPH MONITORING'
+				sig, err := s.(signer.DataSigner).SignData(MonitoringInputData, s.(signer.DataSigner).GetDefaultOptions())
+				if err != nil {
+					sigerrstrs[i] = fmt.Sprintf("signing failed with error: %v", err)
+					return
+				}
+
+				encodedsig, err := sig.Marshal()
+				if err != nil {
+					sigerrstrs[i] = fmt.Sprintf("encoding failed with error: %v", err)
+					return
+				}
+				sigerrstrs[i] = ""
 				sigresps[i] = formats.SignatureResponse{
 					Ref:        id(),
 					Type:       s.Config().Type,
 					Mode:       s.Config().Mode,
 					SignerID:   s.Config().ID,
+					PublicKey:  s.Config().PublicKey,
+					Signature:  encodedsig,
+					X5U:        s.Config().X5U,
 					SignerOpts: s.Config().SignerOpts,
 				}
 				return
 			}
 
-			if _, ok := s.(signer.DataSigner); !ok {
-				sigerrstrs[i] = fmt.Sprintf("signer %q does not implement the DataSigner interface", s.Config().ID)
-				return
-			}
-
-			// base64 of the string 'AUTOGRAPH MONITORING'
-			sig, err := s.(signer.DataSigner).SignData(MonitoringInputData, s.(signer.DataSigner).GetDefaultOptions())
-			if err != nil {
-				sigerrstrs[i] = fmt.Sprintf("signing failed with error: %v", err)
-				return
-			}
-
-			encodedsig, err := sig.Marshal()
-			if err != nil {
-				sigerrstrs[i] = fmt.Sprintf("encoding failed with error: %v", err)
+			if _, ok := s.(signer.FileSigner); ok {
+				if _, ok := s.(signer.TestFileGetter); !ok {
+					sigerrstrs[i] = fmt.Sprintf("signer %q implements FileSigner but not the TestFileGetter interface", s.Config().ID)
+					return
+				}
+				output, err := s.(signer.FileSigner).SignFile(s.(signer.TestFileGetter).GetTestFile(), s.(signer.FileSigner).GetDefaultOptions())
+				if err != nil {
+					sigerrstrs[i] = fmt.Sprintf("signing failed with error: %v", err)
+					return
+				}
+				signedfile := base64.StdEncoding.EncodeToString(output)
+				sigerrstrs[i] = ""
+				sigresps[i] = formats.SignatureResponse{
+					Ref:        id(),
+					Type:       s.Config().Type,
+					Mode:       s.Config().Mode,
+					SignerID:   s.Config().ID,
+					PublicKey:  s.Config().PublicKey,
+					SignedFile: signedfile,
+					X5U:        s.Config().X5U,
+					SignerOpts: s.Config().SignerOpts,
+				}
```

Comment:
@g-k This is a new approach to monitor signers that only implement the FileSigner interface. It adds a new `TestFileGetter` that the signer implements to return a valid test file, in this case it's a pruned version of Focus.

We still don't have a way to verify the signed file (that would mean shelling out to apksigner in `autograph-monitor`), but at least this code exercises the signer and makes sure its configuration is good.

Reply:
WFM 

Long term, do we still want to switch the monitor to fargate (or having a separate file signer verifier) and shell out to `apksigner verify`? If so, I can make an issue for that.

Reply:
I think that would be worthwhile. And perhaps we could also integrate the work mgoodwin is doing around xpcshell in the monitor runtime.